src/ucp/stream/stream_recv.c

```diff
@@ -253,47 +240,36 @@ UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_stream_recv_nb,
 
     UCP_THREAD_CS_ENTER_CONDITIONAL(&ep->worker->mt_lock);
 
-    req = ucp_stream_recv_request_get(ep->worker, buffer, count, datatype, cb);
-    if (ucs_unlikely(req == NULL)) {
-        req = UCS_STATUS_PTR(UCS_ERR_NO_MEMORY);
-        goto out;
-    }
-
-    ep_stream = ep->ext.stream;
-
-    /* if there are already posted requests, need to queue this one */
-    if (ep_stream->reqs_buf_len > 0) {
-        ucs_assert(!ucs_queue_is_empty(&ep_stream->reqs));
+    if (ucs_unlikely(ep_stream->am_data_len == 0)) {
+        /* No data, enqueue request and out */
+        req = ucp_request_get(ep->worker);
+        if (ucs_unlikely(req == NULL)) {
+            req = UCS_STATUS_PTR(UCS_ERR_NO_MEMORY);
+            goto out;
+        }
+        ucp_stream_recv_request_init(req, buffer, count, datatype, cb);
         ucs_queue_push(&ep_stream->reqs, &req->recv.queue);
-        ep_stream->reqs_buf_len += req->recv.length;
         goto ptr_out;
     }
 
-    /* OK, lets obtain all arrived data which matches the recv size */
-    while (req->recv.state.offset < req->recv.length) {
-        rdesc = ucp_stream_recv_data_nb_internal(ep_stream, 0);
-        if (rdesc == NULL) {
-            /* No data any more. Enqueue empty request or complete if some
-             * data is filled */
-            if (req->recv.state.offset == 0) {
-                ucs_queue_push(&ep_stream->reqs, &req->recv.queue);
-                ep_stream->reqs_buf_len += req->recv.length;
-                goto ptr_out;
-            } else {
-                req = ucp_stream_recv_request_put(req, length, UCS_OK);
-                goto out;
-            }
-        }
+    /* We can complete request in place, use tmp request on stack */
```

Comment:
Is there any way to avoid using request when data is available immediately?

Reply:
sure, but it will significantly duplicate and complicate the code, especially for non-contiguous datatypes.

Reply:
ok